kaggler/online_model/ftrl.pyx

```diff
@@ -81,24 +82,27 @@ cdef class FTRL:
             self.a, self.b, self.l1, self.l2, self.n, self.epoch, self.interaction
         )
 
-    def _indices(self, list x):
-        cdef unsigned int index
-        cdef int l
+    cdef list _indices(self, int[:] x):
+        cdef int x_len = x.shape[0]
+        cdef int index
         cdef int i
         cdef int j
+        cdef list indices = []
+        indices.append(self.n)
 
-        # return the index of the bias term
-        yield self.n
-
-        for index in x:
-            yield abs(hash(index)) % self.n
+        for i in range(x_len):
+            index = x[i]
+            indices.append(abs(hash(index)) % self.n)
+            # a much faster and also reasonable way:
+            # indices.append(index % self.n)
 
         if self.interaction:
-            l = len(x)
-            x = sorted(x)
-            for i in xrange(l):
-                for j in xrange(i + 1, l):
-                    yield abs(hash('{}_{}'.format(x[i], x[j]))) % self.n
+            for i in range(x_len):
+                for j in range(i + 1, x_len):
+                    indices.append(abs(hash('{}_{}'.format(x[i], x[j]))) % self.n)
+                    # a much faster and also reasonable way:
+                    # indices.append((x[i] * x[j]) % self.n)
```

Comment:
With this, there are collisions for low indices. e.g. when `x[i] == 1` and `x[j] == 2`, `(x[i] * x[j]) % self.n == 2`.  Alternatively, we may look for a faster `hash` function.